lib/sqlite.c

```diff
@@ -29,18 +29,57 @@
 #include <time.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <libgen.h>
 #include <string.h>
 #include <limits.h>
 #include <sys/stat.h>
 #include <sqlite3.h>
 
 #include "wtmpdb.h"
 
+/* Begin - local helper functions */
+
+static void
+mkdir_p(const char *pathname, mode_t mode)
+{
+  if (mkdir(pathname, mode) == 0 || errno == EEXIST || errno != ENOENT)
+    return;
+
+  char *buf = strdup(pathname);
+  mkdir_p(dirname(buf), mode);
+  free(buf);
+
+  mkdir(pathname, mode);
+}
+
+static void strip_extension(char *in_str)
+{
+    static const int name_min_len = 1;
+    static const int max_ext_len = 4;
+
+    /* Check chars starting at end of string to find last '.' */
+    for (ssize_t i = strlen(in_str); i > (name_min_len + max_ext_len); i--)
+    {
+        if (in_str[i] == '.')
+        {
+            in_str[i] = '\0';
+            return;
+        }
+    }
+}
+
+/* End - local helper functions */
+
+
 static sqlite3 *
 open_database_ro (const char *path, char **error)
 {
   sqlite3 *db;
 
+  char *buf = strdup(path);
+  mkdir_p(dirname(buf), S_IRWXU);
+  free(buf);
+
```

Comment:
Shouldn't that be part of open_database_rw()? In open_database_ro() we will never create a database/write to a database, and open_database_rw() could still be called without calling open_database_ro() before.